### PR TITLE
[lspext] Add the lsp extension - 'edefinition'.

### PR DESCRIPTION
**Pull request by movie-travel-code**
The extension 'definition' needs to compute qname, package locator and
symbol kind.

The 'packages.Load()' method can't resolve the packages located
outside the '$GOPATH/src', except the path is not the relative path,
sourcegraph and the vscode-go both have this problem.

### CODE REVIEW

File: internal/lsp/server.go [hunk 1]
```diff
@@ -369,6 +370,54 @@ func (s *server) Definition(ctx context.Context, params *protocol.TextDocumentPo
 	return []protocol.Location{toProtocolLocation(s.view.FileSet(), ident.Declaration.Range)}, nil
 }
 
+// EDefinition has almost the same functionality with Definition except for the
+// qualified name and symbol kind.
+func (s *server) EDefinition(ctx context.Context, params *protocol.TextDocumentPositionParams) ([]protocol.SymbolLocator, error) {
```
Comment (zfy0701) [inline internal/lsp/server.go]: could you add a test case as well?

File: internal/lsp/server.go [hunk 2]
```diff
@@ -369,6 +370,54 @@ func (s *server) Definition(ctx context.Context, params *protocol.TextDocumentPo
 	return []protocol.Location{toProtocolLocation(s.view.FileSet(), ident.Declaration.Range)}, nil
 }
 
+// EDefinition has almost the same functionality with Definition except for the
```
Comment (zfy0701) [inline internal/lsp/server.go]: is it possible to create a file like

elastic/lsp/server.go and inherit this function?

this way we could avoid potential merge conflict as much as possible

File: internal/lsp/server.go [hunk 3]
```diff
@@ -369,6 +370,54 @@ func (s *server) Definition(ctx context.Context, params *protocol.TextDocumentPo
...
+
+	qname := source.QName(f, ident)
+
+	pkg := f.GetPackage()
+	if pkg == nil {
+		return nil, fmt.Errorf("no packages found for this file")
+	}
+
+	pkgLoc := protocol.PackageLocator{Name: f.GetPackage().Name, RepoURI: string(sourceURI)}
+	loc := toProtocolLocation(s.view.FileSet(), ident.Declaration.Range)
```
Comment (Poytr1) [inline internal/lsp/server.go]: We can check if `loc` is null first, if it isn't null, we may needn't get qname and kind any more.

File: internal/lsp/server.go (hunk 2 above)
Comment (movie-travel-code) [reply internal/lsp/server.go]: done

File: internal/lsp/server.go (hunk 3 above)
Comment (movie-travel-code) [reply internal/lsp/server.go]: Good point! However, IMHO it is a not good time to do this check. It's better to wait until the package dependence resolve finished.

**Comment by movie-travel-code** [comment]
There are several issues that have not been fixed.

- **Bug**: Official lsp can't get the import package list for now, if we try to resolve the symbol imported from other packages, the import packages list is necessary. I will dive into this bug later.
- **Bug**: The test framework `packagestest` can't get the identifier which imported from other packages.
- **Bug**: If the project is not located in `$GOPATH/src`, the package resolve will fail.
- The role of `PackageLocator`is not yet clear, especially the field `RepoURI`. For now it is just a relative path like `golang.org/x/tools/internal/lsp/lspext` or `github.com/...`

And for anonymous name, I ignore it for the time being. Like the code snippets below, the qualified name of x is `main.Circle.x`. Another choice is `main.Circle.Node.x`, which better?
```go
package main
type Circle struct {
  r int
  Node struct {
    x int
    y int 
  }
}
```

**Comment by zfy0701** [comment]
great job, let's do some final cleanup and checkin our first version

> There are several issues that have not been fixed.
> 
> * **Bug**: Official lsp can't get the import package list for now, if we try to resolve the symbol imported from other packages, the import packages list is necessary. I will dive into this bug later.
> * **Bug**: The test framework `packagestest` can't get the identifier which imported from other packages.
> * **Bug**: If the project is not located in `$GOPATH/src`, the package resolve will fail.
one thing I'm thinking about is that we could do symlink of our workspace to a reasonable $GOPATH/src, we could discuss more

> * The role of `PackageLocator`is not yet clear, especially the field `RepoURI`. For now it is just a relative path like `golang.org/x/tools/internal/lsp/lspext` or `github.com/...`

it's alright, let's fix them when we begin to do cross repo jump to definition

> 
> And for anonymous name, I ignore it for the time being. Like the code snippets below, the qualified name of x is `main.Circle.x`. Another choice is `main.Circle.Node.x`, which better?

`main.Circle.Node.x` is better
> 
> ```go
> package main
> type Circle struct {
>   r int
>   Node struct {
>     x int
>     y int 
>   }
> }
> ```

File: internal/lsp/protocol/server.go [hunk 4]
```diff
@@ -228,6 +229,14 @@ func serverHandler(server Server) jsonrpc2.Handler {
 			}
 			resp, err := server.Definition(ctx, &params)
 			unhandledError(conn.Reply(ctx, r, resp, err))
+		case "textDocument/edefinition":
```
Comment (zfy0701) [inline internal/lsp/protocol/server.go]: so now we could get rid of this, is it?

Comment (movie-travel-code) [reply internal/lsp/protocol/server.go]: This case is necessary, otherwise the jsonrpc2 [handle()](https://github.com/golang/tools/blob/master/internal/jsonrpc2/jsonrpc2.go#L327), i.e. `serverHandler` won't recognize the `textDocument/edefinition` request.

**Comment by movie-travel-code** [comment]
> `main.Circle.Node.x` is better
> 
> > ```go
> > package main
> > type Circle struct {
> >   r int
> >   Node struct {
> >     x int
> >     y int 
> >   }
> > }
> > ```
Done.

File: internal/lsp/protocol/server.go (hunk 4 above)
Comment (zfy0701) [reply internal/lsp/protocol/server.go]: I see, is it possible to override `serverHandler`

if it is too much effort we could do that later

if possible then the way I assume we code is something like

switch ()
  case 'edefintion': our logic
  default: super.switch

Comment (movie-travel-code) [reply internal/lsp/protocol/server.go]: Maybe I didn't get what you mean. The `serveHandler` is a trivial function, there is no way to override this function to achieve your expectations. The only approach is to duplicate this function.

**TL;DR** jump to https://drive.google.com/drive/folders/1s9Vmjo581S9RJWayuIXQz78O9U2YBg-5 to get an intuitive understanding about the difference between **approach1** and **approach2**.

I reconsider the advice you provided last week, the idea is **minimize the interference with original lsp as much as possible through the code duplication**. The initial trial, creating a new repo to depend on the lsp, failed because the key functions and structs are not for export.

The main idea of previous version patch, named **approach1**, is hijacking the `lsp.RunServer()` to use `lsp.elasticserver` to replace the `lsp.server`. And modify the `serverHandler()` and `protocol.Server`.

The main idea of current version patch, named **approach2**, is hijacking the `serve.Run()`, earlier than **approach1**, to call the `lsp.RunElasticServer()`(added this time) instead of `lsp.RunServer()`. And create `elasticServerHandler()` copied from `serverHandler`, `protocol.ElasticServer` inherit from `protocol.ElasticServer`.  All in all, reduce the interference as much as possible through the code duplication.

Comment (zfy0701) [reply internal/lsp/protocol/server.go]: sorry, I mis read the code, 

I'm ok with both solution,
could you try the server with Kibana, specifically go to definition and find reference inside repo, and give a demo on Monday?

Comment (movie-travel-code) [reply internal/lsp/protocol/server.go]: sure, I will try it.
